### PR TITLE
add cmd support (-c, --command)

### PR DESCRIPTION
**Pull request by lastquestion**
I want to add a test for this, but because it spans the whole batch script -> python thing, I want some guidance.

Passes pep8.

### CODE REVIEW

**Comment by lastquestion** [comment]
One thing: in `fpp`: I'd like to remove the matching of `record` and do that in `choose.py`, but by doing so, we'd slightly change behavior: we wouldn't print the `record` text until *after* `processInput` finishes. @pcottle: are you fine with that? IMHO it would clean up the code for longer term. Or, we can instantiate a copy of `ScreenFlags` in `processInput` and do the outputting there, if you'd like.

**Comment by lastquestion** [comment]
Another thing: I wanted to support prefilling in the command in *command-mode*, but it was very difficult to do so. Basically, there isn't a default built in text-box in `ncurses`, and the only thing that exists that sort of does what we would want is `urwid` (https://github.com/wardi/urwid), but it's way too heavyweight and the text box there still doesn't support `insert-mode`.

I think we should code up our own text entry widget and turn back on the cursor for the command mode if we want to support prefilled in commands from the command line in that screen. I like the idea of it and we should do it in a separate, follow up PR.

File: src/__tests__/screenTestRunner.py [hunk 1]
```diff
@@ -29,19 +32,21 @@ def getLineObjsFromFile(inputFile):
...
+        printScreen=True,
+        pastScreen=None):
     lineObjs = getLineObjsFromFile(inputFile)
     screen = ScreenForTest(
         charInputs,
         maxX=screenConfig.get('maxX', 80),
         maxY=screenConfig.get('maxY', 30),
     )
+    # mock our flags with an empty command line for now
+    flags = ScreenFlags.initFromArgs([])
```
Comment (pcottle) [inline src/__tests__/screenTestRunner.py]: ah i didnt realize it was so easy to mock arg-based parsing... aka i didn't realize `argparse` could take in an empty array and not complain

File: src/choose.py [hunk 2]
```diff
@@ -91,4 +91,8 @@ def setSelectionsFromPickle(selectionPath, lineObjs):
         output.writeToFile('echo ":D"')
         sys.exit(0)
     output.clearFile()
-    curses.wrapper(doProgram)
+    # we initialize our args *before* we curse up
+    # so we can benefit from the default argparse
+    # behavior:
+    flags = ScreenFlags.initFromArgs(sys.argv)
+    curses.wrapper(functools.partial(doProgram, flags=flags))
```
Comment (pcottle) [inline src/choose.py]: you can also do:

```
lambda x:doProgram(x, flags)
```

do avoid the partial bind but nbd

Comment (lastquestion) [inline src/choose.py]: that's cleaner, I was afraid because I didn't know what python level supported what.

**Comment by lastquestion** [comment]
I'll rewrite to use the lambda and then it should be good to merge?

**Comment by lastquestion** [comment]
also what's your thought about removing the match of `record` from `fpp` and moving it into `choose`

**Comment by pcottle** [comment]
> I'll rewrite to use the lambda and then it should be good to merge?

yeah! i can do it as a followup if you want

> also what's your thought about removing the match of record from fpp and moving it into choose

sure thats fine, i kinda just had it there more for documentation anyways

**Comment by pcottle** [comment]
actually ill just do the followups, want to get my hands on some code today anyways

**Comment by pcottle** [comment]
Hrm @lastquestion i think the issue here is that bash includes the script dir as the first arg, and then python gets that by accident:

```
[pcottle:~/Dropbox (Facebook)/wip/PathPicker:master]$ ./fpp
the opt OPTS
Using old result...
i got args ['/Users/pcottle/Dropbox (Facebook)/wip/PathPicker/src/choose.py']
```

**Comment by pcottle** [comment]
Alright I fixed that but still get errors when using other args:

[pcottle:~/Dropbox (Facebook)/wip/PathPicker:master]$ ./fpp --python3
Using old result...
usage: fpp [-h] [-r] [-ko] [-c COMMAND [COMMAND ...]]
fpp: error: unrecognized arguments: --python3

wonder if we can make it more permissive

**Comment by pcottle** [comment]
think we had another bug:

```
     def initFromArgs(argv):
-        args = ScreenFlags.getArgParser().parse_known_args(argv)
+        (args, chars) = ScreenFlags.getArgParser().parse_known_args(argv)
         return ScreenFlags(args)
```

File: fpp [hunk 3]
```diff
@@ -21,20 +21,22 @@ done
...
+
+  $PYTHONCMD "$BASEDIR/src/choose.py" $OPTS < /dev/tty
   # execute the output bash script
   sh ~/.fpp/.fpp.sh < /dev/tty
 }
 
 # we need to handle the --help option outside the python
 # flow since otherwise we will move into input selection...
-for opt in "$@"; do
+for opt in OPTS; do
```
Comment (pcottle) [inline fpp]: hrm i think this broken the argument parsing in bash, reverting

**Comment by lastquestion** [comment]
Argh, yeah, revert this.

a) I forgot that we are not stripping the other arguments into `argparse`
b) there is something super weird with how "$@" is expanded in bash. 

Sorry, did you revert it already? to fix it, we  should add the other args to the argparse arguments so it doesn't blow up on --python3 etc.

for the other, I think the OPTS expands differently right? I didn't test with --python3 or --ko with -c at the same time for example. My bad again.

**Comment by pcottle** [comment]
no need to revert -- i fixed everything in the commands above. i think everything works now! try fetching from upstream and let me know

**Comment by lastquestion** [comment]
OK, cool. I'll fetch from upstream, double check. Sorry about that again. Is there a good way for us to test this? It's weird cuz it spans the batch to python land.

**Comment by pcottle** [comment]
> Sorry about that again. Is there a good way for us to test this? It's weird cuz it spans the batch to python land.

i think ill add a UI test that just mocks the command being passed in and then checks we present the error message in the ui

**Comment by pcottle** [comment]
In the 0.5.7 release, adding PR to homebrew now:
https://github.com/facebook/PathPicker/releases/tag/0.5.7

**Pull request merged**